Documentation/dr dependency manager (#641)
dxatscale/sfpowerscripts: decision records/003-Depenedency-Manager.md

# Dependency Manager 
•	Status: Proposed
•	Issue: #638

## Context and Problem Statement 
As the number of packages increases in large scale projects, with multiple dependencies, updating package versions and dependencies manually is time consuming and takes great care to update dependencies one by one. 

Other solutions encountered were local bash scripts as documented in both the issue https://github.com/Accenture/sfpowerscripts/issues/638 raised and in the discussions has here: https://github.com/Accenture/sfpowerscripts/discussions/468 

## Solution

> The dependency manager should utilise the SemVer incrementation style https://semver.org/ and be of a similar method to the ‘lerna version’ command. It should prompt the user for the version: (major, minor, patch, custom). 

Reviewer: Lerna has a no prompt support as well, where we say Major it updates major for all packages

Author: Updated

Since salesforce does not necissarily follow the semver versioning, we will provide a summary to the bump. 
Eg. Option 1 bump Major - includes incompatible API changes

> We will also not be supporting pre-release versioning at this stage. 

Reviewer: Can you elaborate here? What does pre-release mean?

Author: A pre-release version MAY be denoted by appending a dash and a series of dot separated identifiers immediately following the patch version. Identifiers MUST be comprised of only ASCII alphanumerics and dash [0-9A-Za-z-]. Pre-release versions satisfy but have a lower precedence than the associated normal version. Examples: 1.0.0-alpha, 1.0.0-alpha.1, 1.0.0-0.3.7, 1.0.0-x.7.z.92.

Reviewer: yeah we dont need pre-release support, as salesforce packages dont support it. We need to find a way for build number for source and data packages, say if the user has any non zero build number, do we bring back to zero?

Author: We should check if the number is non-zero and ask if they want it reset to zero if it is

Reviewer: can you add these to the document?

Steps to the command:
For each package in the sfdx-project.json the command will ask 
1. if the package version should be updated
2. if the package version has been updated, would the user like all packages with this dependency updates as well? 

> 3. Would the user like to update the dependencies of this package? (regardless of whether the package version was updated)

Reviewer: I didn't understand this one? Why do we need to do this?

The packageDirectories portion will then be updated with the input given and written to the sfdx-project.json in the root folder with the option to commit the file (but not push to the repo). 

## Decision 

> Development to commence on the solution and be included in the september release

Reviewer: ```suggestion
A new command would be built with the above solution and will be added to sfpowerscripts
```

# Dependency Manager 
•	Status: Proposed
•	Issue: #638

## Context and Problem Statement 
As the number of packages increases in large scale projects, with multiple dependencies, updating package versions and dependencies manually is time consuming and takes great care to update dependencies one by one. 

Other solutions encountered were local bash scripts as documented in both the issue https://github.com/Accenture/sfpowerscripts/issues/638 raised and in the discussions has here: https://github.com/Accenture/sfpowerscripts/discussions/468 

## Solution
The dependency manager should utilise the SemVer incrementation style https://semver.org/ and be of a similar method to the ‘lerna version’ command. It should prompt the user for the version either for individual packages or for all packages in the sfdx-project.json: (major, minor, patch, custom). 

> Since salesforce does not necissarily follow the semver versioning, we will provide a summary to the bump. 

Reviewer: ```suggestion
Since salesforce does not necessarily follow the semver versioning, 
```

> Eg. Option 1 bump Major - includes incompatible API changes

Reviewer: ```suggestion

```
